### PR TITLE
interface/builtin: add qualcomm-ipc-router interface for AF_QIPCRTR socket protocol

### PR DESCRIPTION
**Pull request by tsunghanliu**
This PR is to make snapd support the `AF_QIPCRTR` socket family.

`socket AF_QIPCRTR` (Qualcomm IPC router protocol) is used to communicate
with services provided by other hardware blocks in the system.

Snaps to access these kinds of Qualcomm hardware components have to use this socket protocol.
For example: a qmi client will use this interface to communicate to the compute DSP ipcrtr and qmi services.

### CODE REVIEW

File: interfaces/seccomp/template.go
Comment (anonymouse64) [inline interfaces/seccomp/template.go:483]: we add support for both AF_QIPCRTR and PF_QIPCRTR to the resolver, can we also add PF_QIPCRTR here too?

**Review by anonymouse64 — COMMENTED** [review]
While AF_QIPCRTR and PF_QIPCRTR seem very hardware specific, the comment about the list of default socket types does explain that it is intentional that all socket types are allowed in the default seccomp policy to be mediated by AppArmor. I'm fairly certain that this socket type will still be mediated by AppArmor too, but we should verify this before putting it into the default template.

File: interfaces/seccomp/template.go
Comment (tsunghanliu) [reply interfaces/seccomp/template.go:483]: `seccomp/template.go` doesn't declare any PF_*, and AF_X is generally the same as PF_X.
So I think it's not necessary to declare PF_*?

**Comment by tsunghanliu** [comment]
Thanks Ian.
To support `AF_QIPCRTR` correctly, a new interface is needed. I'm working on it and will update the PR later.

**Comment by tsunghanliu** [comment]
I think it's necessary to have a separate interface for this socket protocol.

File: interfaces/seccomp/template.go
Comment (anonymouse64) [reply interfaces/seccomp/template.go:483]: sure I guess I didn't notice we only use the `AF_` here and not the `PF_` variants, this is fine then

File: interfaces/builtin/qrtr.go
Comment (anonymouse64) [reply interfaces/builtin/qrtr.go:45]: also if this is already in the default template, it does not need to be included again, in which case I think you can just have this interface declare the apparmor bits

Comment (tsunghanliu) [reply interfaces/builtin/qrtr.go:45]: I was referring to `can-bus.go` in which there's a similar comment-out declaration and I thought it's intentional to note the seccomp template already has it.
I'm not sure if I should keep it, but am happy to remove it.

Comment (anonymouse64) [reply interfaces/builtin/qrtr.go:45]: Sure this is fine to leave here in that case

File: interfaces/builtin/qrtr.go
Comment (anonymouse64) [inline interfaces/builtin/qrtr.go:50]: need input from @pedronis on this name, I think it's kinda okay but it's also quite short and not super clear, other options could be:

* `qualcomm-router-ipc`
* `qrtr-socket`
* `qualcomm-router`
* `qualcomm-router-socket`

**Review by anonymouse64 — APPROVED** [review]
lgtm, but needs input on the name of the interface from others

File: interfaces/builtin/qrtr.go
Comment (pedronis) [inline interfaces/builtin/qrtr.go:33]: this is usually is the same a the description, we can have a secondardy about the sock_dgram only bit

File: interfaces/builtin/qrtr.go
Comment (pedronis) [inline interfaces/builtin/qrtr.go:22]: it would be good to a have a comment with a link to the relevant upstream kernel bits

**Review by pedronis — COMMENTED** [review]
commented on the name plus some smaller details, I also notice that the tests are failing atm because of the missing spread tests

**Review by anonymouse64 — APPROVED** [review]
+1 lgtm

**Comment by tsunghanliu** [comment]
The interface name is changed to `qualcomm-ipc-router`, descriptions and comments are updated and the spread test is added.

**Review by alexmurray — COMMENTED** [review]
In general this LGTM, however one big problem I noticed is that the spread test for this interface is failing on Ubuntu 16.04:

```
 + snap connect test-snapd-policy-app-consumer:qualcomm-ipc-router :qualcomm-ipc-router
error: cannot perform the following tasks:
- Connect test-snapd-policy-app-consumer:qualcomm-ipc-router to core:qualcomm-ipc-router (cannot setup profiles for snap "test-snapd-policy-app-consumer": cannot load apparmor profiles: exit status 1
apparmor_parser output:
AppArmor parser error for /var/lib/snapd/apparmor/profiles/snap.test-snapd-policy-app-consumer.qualcomm-ipc-router in /var/lib/snapd/apparmor/profiles/snap.test-snapd-policy-app-consumer.qualcomm-ipc-router at line 567: Invalid network entry.
```

This is likely because the version of apparmor used there is too old to be aware of this socket type (this was only added in apparmor 2.12 but xenial has apparmor 2.10) - as such, any snap which uses this interface will not be able to be installed on any of these older systems. 

@pedronis et al is it acceptable to add such an interface which will fail hard on older systems like this?

**Comment by mardy** [comment]
I see that the kernel module has an option to expose a `/dev/qrtr-tun` device to allow userspace to control the router through it. Maybe it's also an option we want to enable in the apparmor profile?

**Comment by anonymouse64** [comment]
@alexmurray good catch! What we normally do for this is not omit rules that aren't supported there or otherwise treat that apparmor feature as required for snapd, so if apparmor doesn't support that feature then the whole system would be marked degraded and run everything in devmode.

We probably don't want xenial to be degraded however and AIUI this is primarily intended to be consumed from UC20, so maybe it's good enough to just omit the failing rule on xenial (and other systems where apparmor doesn't support this network type)

Alternatively, we could ask someone from the security team to explore what it would take to backport that feature to xenial. I don't know how much work that would be.

In any case I'm marking blocked until Samuele is back to decide definitively.

**Comment by tsunghanliu** [comment]
> I see that the kernel module has an option to expose a `/dev/qrtr-tun` device to allow userspace to control the router through it. Maybe it's also an option we want to enable in the apparmor profile?

I don't have a use case of `/dev/qrtr-tun`. According to the [config option](https://cateee.net/lkddb/web-lkddb/QRTR_TUN.html), the `dev/qrtr-tun`  allows user space to implement endpoints of QRTR, for purpose of tunneling data to other hosts or testing purposes. It seems to me a different type of usage than the `AF_QIPCRTR`.

I think it's fine to not be included in the initial version of support?

**Comment by tsunghanliu** [comment]
> This is likely because the version of apparmor used there is too old to be aware of this socket type (this was only added in apparmor 2.12 but xenial has apparmor 2.10) - as such, any snap which uses this interface will not be able to be installed on any of these older systems.

@alexmurray is AppArmor SRU (to enable `AF_QIPCRTR` support) a possible solution?

**Comment by mardy** [comment]
> I don't have a use case of `/dev/qrtr-tun`. According to the [config option](https://cateee.net/lkddb/web-lkddb/QRTR_TUN.html), the `dev/qrtr-tun` allows user space to implement endpoints of QRTR, for purpose of tunneling data to other hosts or testing purposes. It seems to me a different type of usage than the `AF_QIPCRTR`.
> 
> I think it's fine to not support `qrtr-tun` in the initial version?

Sure it is :-)

**Comment by anonymouse64** [comment]
@pedronis I pushed a minimal patch which just adds an apparmor parser feature around the specific network rule here and then the interface during AppArmorConnectedPlug() will will query the parser features and check if this rule is supported or not.

I wanted to try and pass in the features through *apparmor.Specification, but I couldn't find anywhere when we create the spec that we would also logically have access to the sandbox already and it seemed a bit of a scope creep to add apparmor specific stuff there, but maybe it's fine to just randomly add apparmor_sandbox.ParserFeatures() somewhere rather than use it directly in the interface, I'm not sure. It feels like the "right" "best" place is somewhere in backend.go, but I'm not sure where that we could set something that communicated down into the spec.

Also there will probably be spread tests that need adjusting to check for this, I didn't look to see what spread tests need to be adjusted.

**Review by mardy — DISMISSED** [review]
IMHO the apparmor feature check is so compact that it doesn't hurt readability or pose any difficulty for future refactoring, so +1 for it.

**Comment by anonymouse64** [comment]
Okay now this actually does the nice thing where the interface queries the apparmor.Specification for the available features in the sandbox which is cleaner and reads more nicely. 

I assume I will still have to adjust a spread test or two, need to see which ones though.

File: interfaces/apparmor/spec.go
Comment (pedronis) [inline interfaces/apparmor/spec.go:108]: this probably can just be called "MockFeatures" no? I don't think we use the word combination "mock set" anywhere so far

**Review by pedronis — APPROVED** [review]
+1 with a naming nitpick

**Comment by pedronis** [comment]
@alexmurray  this should be ready if you want to give it another look

File: interfaces/apparmor/spec.go
Comment (anonymouse64) [reply interfaces/apparmor/spec.go]: done

File: sandbox/apparmor/apparmor.go
Comment (anonymouse64) [reply sandbox/apparmor/apparmor.go:328]: Well, this won't be considered a "user-visible" API just since it's exposed via `snap debug`, since `snap debug` is a debug command and we don't guarantee backwards compatibility of debug commands.

File: sandbox/apparmor/apparmor.go
Comment (anonymouse64) [reply sandbox/apparmor/apparmor.go:327]: done

**Comment by mvo5** [comment]
There is something funky going on in the 16.04 test, i.e. from the log below it looks like it tries to connect the qualcom-ipc interface, fails as expected but then when the next interface is connected it fails again for the qualcom-ipc even if now raw-input is connect.
```
...
2021-09-02T12:38:42.9212825Z + grep -E -q ':qualcomm-ipc-router +'
2021-09-02T12:38:42.9213673Z + snap interfaces
2021-09-02T12:38:42.9215894Z + '[' test-snapd-policy-app-consumer:qualcomm-ipc-router = test-snapd-policy-app-consumer:qualcomm-ipc-router ']'
2021-09-02T12:38:42.9217970Z + os.query is-xenial
2021-09-02T12:38:42.9219271Z + MATCH 'cannot connect plug on system without qipcrtr socket support'
2021-09-02T12:38:42.9221333Z + snap connect test-snapd-policy-app-consumer:qualcomm-ipc-router :qualcomm-ipc-router
2021-09-02T12:38:42.9222681Z + continue
2021-09-02T12:38:42.9223777Z + for plugcmd in '"$SNAP_MOUNT_DIR"/bin/"$CONSUMER_SNAP".*'
2021-09-02T12:38:42.9224941Z + os.query is-debian-10
2021-09-02T12:38:42.9226998Z ++ basename /snap/bin/test-snapd-policy-app-consumer.raw-input
2021-09-02T12:38:42.9229263Z + plugcmd_bn=test-snapd-policy-app-consumer.raw-input
2021-09-02T12:38:42.9230389Z ++ tr . :
2021-09-02T12:38:42.9231852Z ++ echo test-snapd-policy-app-consumer.raw-input
2021-09-02T12:38:42.9233818Z + plug_iface=test-snapd-policy-app-consumer:raw-input
2021-09-02T12:38:42.9235524Z ++ sed s/test-snapd-policy-app-consumer//
2021-09-02T12:38:42.9237247Z ++ echo test-snapd-policy-app-consumer:raw-input
2021-09-02T12:38:42.9238592Z + slot_iface=:raw-input
2021-09-02T12:38:42.9240748Z + '[' test-snapd-policy-app-consumer:raw-input = test-snapd-policy-app-consumer:browser-sandbox ']'
2021-09-02T12:38:42.9243156Z + CONNECTED_PATTERN=':raw-input +.*test-snapd-policy-app-consumer'
2021-09-02T12:38:42.9244776Z + DISCONNECTED_PATTERN=':raw-input +-'
2021-09-02T12:38:42.9245863Z + grep -E -q ':raw-input +'
2021-09-02T12:38:42.9247397Z + snap interfaces
2021-09-02T12:38:42.9249544Z + '[' test-snapd-policy-app-consumer:raw-input = test-snapd-policy-app-consumer:qualcomm-ipc-router ']'
2021-09-02T12:38:42.9252629Z + echo 'When slot :raw-input is connected'
2021-09-02T12:38:42.9253820Z When slot :raw-input is connected
2021-09-02T12:38:42.9254866Z + grep -E -q ':raw-input +-'
2021-09-02T12:38:42.9255524Z + snap interfaces
2021-09-02T12:38:42.9266961Z + '[' :raw-input = :broadcom-asic-control ']'
2021-09-02T12:38:42.9268560Z + '[' :raw-input = :firewall-control ']'
2021-09-02T12:38:42.9269864Z + '[' :raw-input = :kubernetes-support ']'
2021-09-02T12:38:42.9271160Z + '[' :raw-input = :microstack-support ']'
2021-09-02T12:38:42.9272488Z + '[' :raw-input = :openvswitch-support ']'
2021-09-02T12:38:42.9273610Z + '[' :raw-input = :ppp ']'
2021-09-02T12:38:42.9275509Z + snap connect test-snapd-policy-app-consumer:raw-input :raw-input
2021-09-02T12:38:42.9276826Z error: cannot perform the following tasks:
2021-09-02T12:38:42.9280536Z - Connect test-snapd-policy-app-consumer:raw-input to core:raw-input (cannot setup profiles for snap "test-snapd-policy-app-consumer": cannot obtain apparmor specification for snap "test-snapd-policy-app-consumer": cannot connect plug on system without qipcrtr socket support)
2021-09-02T12:38:42.9283542Z -----
2021-09-02T12:38:42.9283993Z .
2021-09-02T12:38:42.9285516Z 2021-09-02 12:38:42 Preparing google:ubuntu-16.04-64:tests/main/install-sideload:reexec0 (sep021222-469961)...
2021-09-02T12:38:43.2252344Z 2021-09-02 12:38:43 Debug output for google:ubuntu-16.04-64:tests/main/interfaces-many-core-provided (sep021222-469987) : 
...
```

**Comment by anonymouse64** [comment]
So I can reproduce the issue with xenial and being unable to connect any other interface after attempting a connection of qualcomm-ipc-router on xenial (and failing), the issue is that we start to perform the connection in repository.Connect() and if that succeeds, by the time we get to AppArmorConnectedPlug() inside SnapSpecification(), we already have saved the connection inside the Repository which the backend uses, so any further attempts to modify the profile will keep trying to generate the profile using that connection and robustly fails every time. What we would probably need to do in order to keep the logic to skip generation for xenial with this socket is to discard connections that result in a failure during Setup() or SetupMany(), but that change seemed non-trivial to me.

As a temporary work-around I have switched back to rejecting the connection in BeforeConnectPlug (though the BeforePreparePlug would work too - both of these methods work) and this is successful in failing the connection such that it does not persist in the interface Repository for the next connection of a different interface, but we are then back to having this interface poke the sandbox directly instead of going through the backend. It's unclear to me if we could instead refactor Connect() somehow to have the Specification available to it, but it doesn't seem like that's much of an option to me since we explicitly only create the Specification after the Connection has been created.

**Comment by anonymouse64** [comment]
Ah actually unit testing BeforeConnectPlug is slightly annoying, so I'm going to use BeforePreparePlug instead

**Comment by anonymouse64** [comment]
I also reverted the additional changes to apparmor.Specification since we don't use it here anymore and it's unclear if we want to keep that code if it can't be used for this singular use case and more generally since the kind of decision we want to make with the features cannot really be made from the Connect stage where we do have the Specification available

**Comment by pedronis** [comment]
We should remember  that BeforePreparePlug is used early when we sanitize
the plugs even if they are not going to be connected, so that’s not ideal
either.

**Comment by anonymouse64** [comment]
@pedronis I suppose in that case we need to do BeforeConnectPlug then, oh well I suppose I can do the testing bits too

**Comment by anonymouse64** [comment]
ok, this has been updated to use BeforeConnectPlug which hopefully should not have any issues, but indeed using BeforePreparePlug caused the interfaces-core-many-provided test to fail on xenial like so:

```
+ snap connect test-snapd-policy-app-consumer:account-control :account-control
WARNING: There is 1 new warning. See 'snap warnings'.
...
+ MATCH 'cannot connect plug on system without qipcrtr socket support'
+ snap connect test-snapd-policy-app-consumer:qualcomm-ipc-router :qualcomm-ipc-router
grep error: pattern not found, got:
error: snap "test-snapd-policy-app-consumer" has no plug named
       "qualcomm-ipc-router"
-----

```

where the warning is about the invalid plug not being able to be prepared

**Comment by mvo5** [comment]
Fwiw, I addressed (hopefully) the original issue with the use of AppArmorConnectedPlug in https://github.com/snapcore/snapd/pull/10728 - I have no strong opinion how we fix it, I'm fine using a workaround here or the fix I proposed I just want this to go in today if at all possible :)

**Review by mardy — APPROVED** [review]
LGTM!

**Comment by anonymouse64** [comment]
.... and one more spread test tweak to include uc16 with xenial in the specific check for this interface ....

and then we should be golden!

**Pull request merged**
